FancyBatching.md

```diff
@@ -0,0 +1,505 @@
+            {
+                AutoFlush = true, // <-- Default
+                AutoFlushInterval = TimeSpan.FromSeconds(5)
+            });
+    await indexer.UploadDocumentsAsync(products);
+    ```
+
+#### I want fine grained control of failure.
+
+    ```C#
+    SearchClient client = ...;
+    await using SearchIndexingBufferedSender<Product> indexer =
+        client.CreateIndexingBufferedSender<Product>();
+    indexer.ActionFailedAsync +=
+        (IndexDocumentsAction<Product> action,
+         IndexingResult result,
+         Exception exception,
+         CancellationToken cancellationToken) =>
+        {
+            // Do work...
+        };
+    await indexer.UploadDocumentsAsync(products);
+    ```
+
+#### I want to checkpoint documents in case my process dies.
+
+    ```C#
+    SearchClient client = ...;
+    await using SearchIndexingBufferedSender<Product> indexer =
+        client.CreateIndexingBufferedSender<Product>();
+    indexer.ActionAddedAsync +=
+        (IndexDocumentsAction<Product> action,
+         CancellationToken cancellationToken) =>
+        {
+            // Save action
+        };
+    indexer.ActionCompletedAsync +=
+        (IndexDocumentsAction<Product> action,
+         IndexingResult result,
+         Exception exception,
+         CancellationToken cancellationToken) =>
+        {
+            // Remove action
+        };
+    await indexer.UploadDocumentsAsync(products);
+    ```
+
+## Obtaining a SearchIndexingBufferedSender
+
+- `SearchClient.CreateIndexingBufferedSender<T>(SearchIndexingBufferedSenderOptions<T>)`
+    - Instantiates a new `SearchIndexingBufferedSender` tied to the current client.
+    - The sender knows about the type of documents in its index
+        - Could also be used with `SearchDocument` for unknown document types
+    - The only way of obtaining a `SearchIndexingBufferedSender` (i.e., no .ctor)
+    - `virtual` as the entry point for mocking fancy batching scenarios
+
+## Configuring SearchIndexingBufferedSender
+
+- `public class SearchIndexingBufferedSenderOptions<T> { ... }`
+
+    - The generic parameter `T` is the type of documents the
+    `SearchIndexingBufferedSender` will manage.  At one point this tyep had to
+    be generic for the events, but now it's only for the `KeyFieldAccessor`
+    which could be simplified to take `object` if we wanted to remove the type
+    parameter here.
+
+    - `bool AutoFlush { get; set; } = true;`
+
+    - `TimeSpan? AutoFlushInterval { get; set; } = TimeSpan.FromSeconds(60);`
+
+    - `CancellationToken AutoFlushIntervalCancellationToken { get; set; }`
+    Specifically for when the auto flush happens from the interval elapsed timer.
+
+    - `Func<T, string> KeyFieldAccessor { get; set; }`  Used to get the key of
+    a document in case we need to associate errors with their inputs.  We'll use
+    this first if provided.
+
+    - `private int? BatchSize  { get; set; }` This will refer to the length of
```

Comment:
Are we moving away from having a batch size indicator for number of documents?

Reply:
I am, yes.  It'll be way more reliable.

Reply:
Let's discuss this with the service team on whether # of documents or size of total payload is better, whether we want to optimize for full payloads or not, what would a good default full payload size look like?

Reply:
Let's also confirm if there's a max of N documents and use that as a hard upper limit with payload size, if we go that route.